c8d/image: Simplify `presentImages` and better "platform not found" error

vvoland opened the pull request:
**- What I did**
A minor refactor, best to review commit by commit.
Summary:
- Simplify `presentImages` to  `getBestPresentImageManifest` - all usages were only using the first sorted result
- Add `errPlatformNotFound` which produces a better error message in case where the image exists, but the requested platform is not present
- Extract `matchRequestedOrDefault` function to refactor a common pattern of creating a platform matcher from `*ocispec.Platform`

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
daemon/c8d: Simplify `presentImages` signature

The `refOrId` parameter is only needed to construct an error in case
where the doesn't have the matching platform.

Move the responsibility of creating the error message to the caller.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d: Improve error message for platform not found

Return a similar error as the graphdrivers implementation when an image
was found, but the requested platform is not present locally or in the
image.

The message doesn't include the "actual" platform, as it doesn't make
sense with the multi-platform images. With graphdrivers all images were
single platform.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d: Simplify `presentImages` into `getBestPresentImageManifest`

All two usages only care about the "first" result from the slice sorted
according to the platform preference.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/history: Unmarshal only a subset of ImageConfig

Only two fields from the whole image config are used, so only unmarshal
these.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
daemon/c8d: Fill wanted platform in GetImageManifest

When `getBestPresentImageManifest` fails with `errPlatformNotFound` -
fill the requested platform.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/image: Add `matchRequestedOrDefault`

Refactor a pattern where a passed `*ocispec.Platform` was used to
create a platform matcher that matches the passed platform if not nil
and uses a default host platform otherwise into a separate function.

Also add some basic unit tests for its behavior.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented on daemon/containerd/image.go:
```diff
@@ -94,8 +97,7 @@ func (i *ImageService) presentImages(ctx context.Context, desc containerdimages.
 	if len(presentImages) == 0 {
-		ref, _ := reference.ParseAnyReference(refOrID)
-		return nil, images.ErrImageDoesNotExist{Ref: ref}
+		return nil, errdefs.NotFound(errors.New("no matching image found"))
```
We may have to be careful with this one; ISTR we even had to change the formatting of the `ErrImageDoesNotExist` error back to use a uppercase "No such image" because some tools depended on that to string-match errors 🙈 https://github.com/moby/moby/blob/bd1d35d72123223b9d4200ab10ce51796bf1967c/daemon/images/image.go#L34

vvoland replied:
It gets transformed back into the ErrImageDoesNotExist by the caller:
https://github.com/moby/moby/blob/aab17154e66c982c58336c98d774f1a8e14e35dc/daemon/containerd/image_history.go#L32-L36

Also, this is replaced by the next commit with a platform not found error.

thaJeztah replied:
Ah, yes, I saw the next commit removed that part. Wondering if we need to be on the safe side and prefix the "platform not found" error with `No such` or `No such image:` (and, yes, I hate it!).

vvoland replied:
I took the message from the graphdriver implementation: https://github.com/moby/moby/blob/bd1d35d72123223b9d4200ab10ce51796bf1967c/daemon/images/image.go#L203 which doesn't have that prefix.

I think it wouldn't hurt though.. WDYT?

thaJeztah replied:
Good call; I didn't check the non-containerd implementation; looking at that one;

- original error was introduced in https://github.com/moby/moby/commit/7a9cb29fb980c0ab3928272cdc24c7089b2fcf64 (https://github.com/moby/moby/pull/40725)
- moved in https://github.com/moby/moby/commit/88c02716053d3cb6f88b17ea9d4a3f21f4d8bcba (https://github.com/moby/moby/pull/41556)
- moved again in https://github.com/moby/moby/commit/50f39e724707ef121a988e422eb52045d3754802 (https://github.com/moby/moby/pull/41955)
    - related to https://github.com/docker/for-linux/issues/1170#issuecomment-750478242

First two were part of 20.10, last one is 23.0, so I think we should be good. We should check if there's no special handling in our own code (e.g. `docker run` using it to check if a pull is needed), and I guess for those string-matching CLI output, we _could_ (worst case scenario) still prepend it in the CLI when printing.

> I think it wouldn't hurt though.. WDYT?

Well, also trying to burn that stuff with fire. If it's proven that for this case we _don't_ need it, we should try to avoid it, otherwise next time we make changes, we need to ask ourselves the same question again ("was this one of those that MUST have the `No such` prefix?")

vvoland replied:
Hmm, in this case I think we can skip adding the prefix. The graphdrivers implementation has been using this message so consumers should already expect it in this form.

laurazard commented on daemon/containerd/image.go:
```diff
@@ -87,116 +90,56 @@ func (i *ImageService) getImageV1(ctx context.Context, img containerdimages.Imag
 
-		return &childManifests[0], nil
+	if best == nil {
+		return nil, &errPlatformNotFound{imageRef: imageFamiliarName(img)}
```
should we not set the wanted platform here?

laurazard replied:
Nevermind, saw that we're doing this in the caller. Wonder if we could move that.

laurazard replied:
I see. I don't love having to remember to do that, but oh well 🤷‍♀️ Maybe we could mention that in `getBestPresentImageManifest`'s godoc?

vvoland replied:
Hmm I got a funny idea, let me try it out 🙈

vvoland replied:
So, I changed `matchRequestedOrDefault` to return a wrapper that also has the requested platform bundled in and use that to populate the error when it's created, so the caller doesn't need to do it.

WDYT?

laurazard replied:
Looks good to me, @thaJeztah?

vvoland committed:
c8d: Wrap matchRequestedOrDefault result with a platform

Allowing it to obtain a specific `ocispec.Platform` used to create this
platform matcher.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented on daemon/containerd/image.go:
```diff
@@ -25,17 +25,46 @@ import (
+
+func (e *errPlatformNotFound) NotFound() {}
+func (e *errPlatformNotFound) Error() string {
+	msg := "image with reference " + e.imageRef + " was found but does not match the specified platform"
```
Perhaps `requested` instead of  `specified` would work for this? Thinking about situations where no explicit platform was requested, so the default was used (or don't we produce the error in any of those cases)?


Also considering if `match` -> `provide` (or `support`); which may be more accurate for multi-arch 🤔

vvoland replied:
It would! But that would be different from the graphdrivers error returned in the same case. So that could break potential error string matching.

IMO, until we provide a better way for consumers to check for this error, we should stick to the existing wording.

thaJeztah replied:
> IMO, until we provide a better way for consumers to check for this error, we should stick to the existing wording.

Hmm.. right; you mean distinction between "image present, but not platform" vs "image not present"? FWIW; I was more concerned about the `No such` errors, because those prefixes were used in docker itself to match error-types; for the errors we're modifying here, we never had that as a contract, so I'm much more OK with changing the error format.

thaJeztah replied:
> but unfortunately Go stringly-typed errors forced that upon us implicitly I think 😅

For exported, sentinel errors, for sure (but also the reason we chose to use the "interface-matching" approach in errdefs). I think in this case, the error itself is never returned to the client as a go-error that can be matched, as the API is in between, so at most the client receives a typed error (errdefs type that is) with an error message that should be considered informational, but in itself not something to control flow.

> but I found at least one instance:

`:sad_panda:` yeah, that's not great. At least it looks like that code already calls it out for being brittle and temporary, and it looks like the issue linked to that comment was closed / addressed https://github.com/buildpacks/pack/issues/2079. That said, I agree we may have to look if the current logic of an image either being "there" or "not there" (`errdefs.NotFound`) is still sufficient to handle logic on the client side, or if we need more specific errors (that also make it accross the API).

I'm ok with keeping the error-message as-is for now, but we should indeed make sure that we don't paint ourself (again) into a corner where error-strings become an undocumented API, and we _must_ preserve error-messages (like we had with the `No such` prefix).

thaJeztah replied:
Perhaps we should add a section to the API docs describing that error-message MUST be considered informational, and can change at any time, so string-matching being highly discouraged.

thaJeztah commented on daemon/containerd/image.go:
```diff
@@ -25,17 +25,46 @@ import (
+	msg := "image with reference " + e.imageRef + " was found but does not match the specified platform"
+	if e.wanted.OS != "" {
+		msg += ": wanted " + platforms.Format(e.wanted)
+	}
```
Playing around a bit as I was wondering if putting this after a `:` delimiter may feel like it's a separate error that was wrapped. I guess we could included it between braces;

```
image with reference foo:bar was found but does not match the requested platform (linux/amd64)
```


Only downside of that would be if we'd have windows (` the requested platform (windows(10.0.17763)/amd64)`)

vvoland replied:
The `wanted: ` part is also present in the graphdrivers error, see above comment.

thaJeztah commented on daemon/containerd/image.go:
```diff
@@ -97,7 +126,9 @@ func (i *ImageService) presentImages(ctx context.Context, img containerdimages.I
-		return nil, errdefs.NotFound(errors.New("no matching image found"))
+		return nil, &errPlatformNotFound{
+			imageRef: img.Name,
+		}
```
Is this situation is only if the manifest index does not provide a usable platform, or also if it does (but the content isn't present)?

vvoland replied:
Both for now.

thaJeztah replied:
To add to the above; the reason I was wondering is because I know we have logic on the client-side where we have special handling for `not found` errors; when doing a `docker run` (or `docker create`), and we get a 404, the CLI initiates a `docker pull` (image not found; let's try pulling it).

Very likely not for this PR, but something we need to start looking at, because with multi-arch there will be more permutations of that;

- The image is missing as a whole; we'll try pulling the image (manifest-index) and the desired platform.
- We pulled the manifest-index, and it has an image for the given platform, but it's not yet pulled. In this case, we can pull the missing platform (but keep the manifest-index at the same version)
- We pulled the manifest-index, but it does not contain the given platform

The last one is the tricky one, at least with the current logic not being granular enough;

- We can either bail out (we have the manifest-index, but unfortunately it doesn't have this platform)
- We can try pulling the manifest-index to see if perhaps a current version of it _does_ have the platform (maybe the image was updated since we pulled the manifest-index?)

Either of the above are "valid", but we don't have a lot of granularity for this currently. The last one also can be confusing, as pulling the manifest-index may mean that any image (for other platform) may now become "dangling" as they no longer match the digest of the newly pulled manifest-index.

thaJeztah commented on daemon/containerd/image_manifest.go:
```diff
@@ -223,12 +223,18 @@ func (m *ImageManifest) ImagePlatform(ctx context.Context) (ocispec.Platform, er
+
+// ReadConfig gets the image config and unmarshals it into the provided struct.
+// The provided struct should be a pointer to the config struct or its subset.
+func (m *ImageManifest) ReadConfig(ctx context.Context, outConfig any) error {
```
Adding `any` may require this file to have a `go:build` comment (https://github.com/moby/moby/pull/46941) (although I don't think this file would be imported by consumers)


Looks like this is only used internally; so we may not need to export it; more so because having to pass the type to read feels a bit "internal"; i.e., if it's encapsulated in the `ImageManifest` type, then somewhat expecting that to know what it returns; being able to unmarshal to something else is mostly an internal optimisation.

vvoland replied:
I can change it to `interface{}`.

This struct already has most of the methods marked as exported (to signify that they're not "internal" and should be used by the consumers of the `ImageManifest`).

I think we could just unexport the whole `ImageManifest` instead.. WDYT?

vvoland replied:
Unexporting the method itself is more of an indicator that the method is "internal" to the `ImageManifest` itself though, which is not the case here as it's used outside of the struct itself by the `ImageService`.

thaJeztah replied:
Maybe. It's just that an interface could still be matched if it would ever be returned, so doesn't really remain internal; e.g. the `errPlatformNotFound` is not exported, but still satisfies the `errdefs.NotFound` interface.

It shouldn't be _really_ problematic (just don't do that!), just thinking if un-exporting it makes it even more clear and it for sure can help looking at code, _knowing_ that it cannot be used outside of the package, without having to go search if it _possibly_ was exported to match some interface .. somewhere..

thaJeztah reviewed (approved):
LGTM
